server/manager/src/main/java/org/apache/accumulo/manager/Manager.java

```diff
@@ -1057,11 +1059,29 @@ private long balanceTablets() {
         int attemptNum = 0;
         do {
           log.debug("Balancing for tables at level {}, times-in-loop: {}", dl, ++attemptNum);
-          params = BalanceParamsImpl.fromThrift(tserverStatusForBalancerLevel,
-              tserverStatusForLevel, partitionedMigrations.get(dl));
+
+          SortedMap<TabletServerId,TServerStatus> statusForBalancerLevel =
+              tserverStatusForBalancerLevel;
+          if (attemptNum % checkInterval == 0) {
```

Comment:
Instead of adding this check, what about instead replacing the do/while loop with a break out of the for loop.  I think the loop is intended to prevent balancing of user tablets when metadata is not balanced.  This also seems to prevent other work that a thread that calls balancing is doing.  This change redoing a subset of that work here.  If the function returned instead of looping it would allow that work to be done.

Wondering if we could do something like the following at this [line](https://github.com/apache/accumulo/blob/8f52dd10f7cbfef192c9cc431543619b7072139a/server/manager/src/main/java/org/apache/accumulo/manager/Manager.java#L1073) instead of the while loop.

```java
if(migrationsOutForLevel > 0 && (dl == DataLevel.ROOT || dl == DataLevel.METADATA)){
   // do not continue balancing later levels until this level is balanced.  
   break;
}
```

Then maybe the outer code/thread will do its thing and eventually call balanceTablets again and maybe get further down the levels.

Reply:
Even though we are passing the balancer a set of tablets for the root or metadata table, that doesn't prevent it from returning migrations for other tables. I think the HostRegexTableLoadBalancer will do this [here](https://github.com/apache/accumulo/blob/2.1/core/src/main/java/org/apache/accumulo/core/spi/balancer/HostRegexTableLoadBalancer.java#L400). In this case, we could be operating on the root or metadata level with `migrationsOutForLevel` being greater than zero. If we exit and call this method again, then I think we end up at the same place.

Reply:
> I think the HostRegexTableLoadBalancer will do this [here](https://github.com/apache/accumulo/blob/2.1/core/src/main/java/org/apache/accumulo/core/spi/balancer/HostRegexTableLoadBalancer.java#L400).

Ok and following that tablet id map it gets it from [here](https://github.com/apache/accumulo/blob/5715f06fc8eba911c89ffa86ee12dff1a64127aa/core/src/main/java/org/apache/accumulo/core/spi/balancer/HostRegexTableLoadBalancer.java#L382) which is coming from this [balancerEnv](https://github.com/apache/accumulo/blob/5715f06fc8eba911c89ffa86ee12dff1a64127aa/server/manager/src/main/java/org/apache/accumulo/manager/Manager.java#L423) that is only initialized once for the manager process.  So when the manager is intending to only balance the root table its not restricting that table id map, so the balancer is completely unaware of this outer goal.

This is making me wonder if we should back out all of the changes for balancing by level in 2.1 for now and try to figure out how to accomplish that goal in a better way.  As it stands now the manager code has a goal for only balancing the root or metadata table and the balancer plugin is completely unaware of that goal.  Going down the path of making the balancer aware of those goals may introduce more bugs in 2.1.  That is why I am wondering if we should back out the changes for now and try to figure out how to balance by level in follow on work

Reply:
It may be that the best way to balance by level in 2.1 is inside a balancer plugin.  Not sure, but maybe doing this at the manager level requires changes to the SPI which would be best done in a 2.2, 3.1, or 4.0.  Not sure though.